valueset.go

```diff
@@ -0,0 +1,267 @@
+	ret := make(map[string]Nodeset)
+	getColumn := func(row []string, header string) string {
+		for cIdx, c := range input.ColumnNames() {
+			if c == header && cIdx < len(row) {
+				return row[cIdx]
+			}
+		}
+		return ""
+	}
+	for {
+		if err := input.Reset(); err != nil {
+			return nil, err
+		}
+		var currNodesetID string
+		var nodeset Nodeset
+		// needs termination condition -- endless loop
+		for {
+			row, err := input.Next()
+			if err == io.EOF {
+				if currNodesetID != "" {
+					ret[currNodesetID] = nodeset
+				} else {
+					processed = false
+				}
+				currNodesetID = ""
+				break
+			}
+			nodesetId := getColumn(row, "nodeset_id")
+			if _, seen := ret[nodesetId]; seen {
+				continue
+			}
+			currNodesetID = nodesetId
+			nodeset = Nodeset{
+				ID:     currNodesetID,
+				Labels: make([]string, 0),
+				Data: map[string]NodesetData{
+					currNodesetID: NodesetData{
+						ID:         currNodesetID,
+						Labels:     make([]string, 0),
+						Properties: make(map[string]interface{}),
+					},
+				},
+			}
+			// // Process a new nodeset, parse row and add properties nodeset
+			parseRow := func(row []string) error {
+				for idx := 0; idx < len(input.ColumnNames()); idx++ {
+					val := row[idx]
+					header := input.ColumnNames()[idx]
+					if header == "node_labels" {
+						if nsl, ok := nodeset.Data[currNodesetID]; ok {
+							removeEnclosures := func(s string, rem ...string) string {
+								for _, r := range rem {
+									s = strings.ReplaceAll(s, r, "")
+								}
+								return s
+							}
+							fields := strings.Fields(row[idx])
+							labels := make([]string, 0)
+							for _, f := range fields {
+								labels = append(labels, removeEnclosures(f, "{", "}", "(", ")", "[", "]", ","))
+							}
+							nsl.Labels = append(nsl.Labels, labels...)
+							nodeset.Data[currNodesetID] = nsl
+						}
+					} else {
+						props, _ := nodeset.Data[currNodesetID].Properties[input.ColumnNames()[idx]].([]string)
+						props = append(props, val)
+						nodeset.Data[currNodesetID].Properties[input.ColumnNames()[idx]] = props
+					}
+				}
+				return nil
+			}
+			if err := parseRow(row); err != nil {
+				return nil, err
+			}
+			processed = true
+			for {
+				r, err := input.Next()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return nil, err
+				}
+				// continue scanning with only with the same nodesetID
+				if currNodesetID != "" && nodesetId != currNodesetID {
+					continue
+				}
+				if err := parseRow(r); err != nil {
+					return nil, err
+				}
+			}
+		}
+		if !processed {
+			break
+		}
+	}
+	return ret, nil
+}
+
+func buildNodesetNodes(nodeset Nodeset) []*lpg.Node {
+	grph := ls.NewDocumentGraph()
+	nodes := make([]*lpg.Node, 0)
+	for _, nsD := range nodeset.Data {
+		nodes = append(nodes, grph.NewNode(nsD.Labels, nsD.Properties))
+	}
+	return nodes
+}
+
+func diff(oldNodeset, newNodeset Nodeset) (insertions Nodeset, deletions Nodeset, updates Nodeset) {
```

Comment:
You'll need some helpers for diff:
  * CompareLabels(oldLabels, newLabels) bool
  * CompareProps(oldProps, newProps) bool
 (these two should be in merge.go)
For each nodeset:
   * Compare root nodes, decide whether to ins, upd, or del
   * For each nodesetData, compare matching, and decide whether to ins, upd, del
  
Diff returns: a flag saying (root: ins, upd, delete, nop), and (insertNodes []NodesetData (these will be from newNodeset), updateNodes []NodesetData (these will be from newNodeset), deleteNodes []string (these will be entityIds of oldNode set))

For each nodesetData in newNodeset, find in oldNodeset. If does not exist, insert that. If exists, compare, either don't do anything, or update.

For each nodesetData in oldNodeset, find in newNodeset, if it does not exist, and if it is not connected anywhere else, delete it.